File: 2015/7/rust/src/main.rs
use std::collections::HashMap;

enum Instruction {
    AND(String, String, String),
    OR(String, String, String),
    NOT(String, String),
    LSHIFT(String, String, String),
    RSHIFT(String, String, String),
    ASSIGN(String, String),
}

fn main() {
    // get command line arguments
    let args: Vec<String> = std::env::args().collect();
    if args.len() < 2 {
        println!("Usage: {} <input.txt>", args[0]);
        return;
    }

    let filename: &str = &args[1];
    let contents = std::fs::read_to_string(filename).expect("Failed to read file");
    let lines = contents.lines();

    let mut instructions: Vec<Instruction> = Vec::new();
    let mut wires: HashMap<String, u16> = HashMap::new();

    for line in lines {
        let parts: Vec<&str> = line.split_whitespace().collect();
        let instruction = match parts.len() {
            3 => Instruction::ASSIGN(parts[0].to_string(), parts[2].to_string()),
            4 => Instruction::NOT(parts[1].to_string(), parts[3].to_string()),
            5 => {
                let op = parts[1];
                let arg1 = parts[0];
                let arg2 = parts[2];
                let dest = parts[4];
                match op {
                    "AND" => Instruction::AND(arg1.to_string(), arg2.to_string(), dest.to_string()),
                    "OR" => Instruction::OR(arg1.to_string(), arg2.to_string(), dest.to_string()),
                    "LSHIFT" => {
                        Instruction::LSHIFT(arg1.to_string(), arg2.to_string(), dest.to_string())
                    }
                    "RSHIFT" => {
                        Instruction::RSHIFT(arg1.to_string(), arg2.to_string(), dest.to_string())
                    }
                    _ => panic!("Unknown operator: {}", op),
                }
            }
            _ => panic!("Invalid instruction: {}", line),
        };

        instructions.push(instruction);
    }

    let a = eval("a", &instructions, &mut wires);
    println!("Part 1: {}", a);
    wires.clear();
    wires.insert("b".to_string(), a);
    let a = eval("a", &instructions, &mut wires);
    println!("Part 2: {}", a);
}

fn eval(target: &str, instructions: &[Instruction], wires: &mut HashMap<String, u16>) -> u16 {
    if let Some(&value) = wires.get(target) {
        return value;
    }

    let instruction = instructions.iter().find(|i| match i {
        Instruction::ASSIGN(_, dest)
        | Instruction::NOT(_, dest)
        | Instruction::AND(_, _, dest)
        | Instruction::OR(_, _, dest)
        | Instruction::LSHIFT(_, _, dest)
        | Instruction::RSHIFT(_, _, dest) => dest == target,
    });

    let instruction = match instruction {
        Some(instruction) => instruction,
        None => panic!("No instruction found for target: {}", target),
    };

    let (op1, op2) = match instruction {
        Instruction::ASSIGN(op1, _) | Instruction::NOT(op1, _) => {
            let op1 = op1
                .parse()
                .unwrap_or_else(|_| eval(op1, instructions, wires));
            (op1, 0)
        }
        Instruction::AND(op1, op2, _)
        | Instruction::OR(op1, op2, _)
        | Instruction::LSHIFT(op1, op2, _)
        | Instruction::RSHIFT(op1, op2, _) => {
            let op1 = op1
                .parse()
                .unwrap_or_else(|_| eval(op1, instructions, wires));
            let op2 = op2
                .parse()
                .unwrap_or_else(|_| eval(op2, instructions, wires));
            (op1, op2)
        }
    };

    let result = match instruction {
        Instruction::ASSIGN(_, _) => op1,
        Instruction::NOT(_, _) => !op1,
        Instruction::AND(_, _, _) => op1 & op2,
        Instruction::OR(_, _, _) => op1 | op2,
        Instruction::LSHIFT(_, _, _) => op1 << op2,
        Instruction::RSHIFT(_, _, _) => op1 >> op2,
    };

    wires.insert(target.to_string(), result);
    result
}

File: 2015/javascript/7/main.js
const fs = require('fs');

class Instruction {
    constructor(type, input1, input2 = "", output){
        this.type = type;
        this.input1 = input1;
        this.input2 = input2;
        this.output = output;
    }

    getInputs(){
        return [this.input1, this.input2];
    }

    getOutput(){
        return this.output;
    }   
}

let instructions = [];
let hashTable = {};

function main(){
    // get command line arguments
    const args = process.argv.slice(1);
    if (args.length < 2){
        console.log("Usage: node main.js <input.txt>");
        return;
    }

    const filename = args[1];

    if (!fs.existsSync(filename)){
        console.log("File not found: " + filename);
        return;
    }

    const data = fs.readFileSync(filename, 'utf8');

    if (!data){
        console.log("File is empty: " + filename);
        return;
    }

    const lines = data.split("\n");


    for(let i = 0; i < lines.length; i++){
        const line = lines[i].trim();
        if (line.length == 0){
            continue;
        }

        const parts = line.split(" ");

        if (parts.length == 3){
            const instruction = new Instruction("ASSIGN", parts[0], "", parts[2]);
            instructions.push(instruction);
        }

        if (parts.length == 4){
            // NOT
            const instruction = new Instruction("NOT", parts[1], "", parts[3]);
            instructions.push(instruction);
        }

        if (parts.length == 5){
            const instruction = new Instruction(parts[1], parts[0], parts[2], parts[4]);
            instructions.push(instruction);
        }
    }

    let a = eval("a");
    console.log("Part 1: " + a);
    hashTable = {};
    hashTable["b"] = a;
    a = eval("a");
    console.log("Part 2: " + a);
}


function eval(target){
    if (hashTable[target] !== undefined){
        return hashTable[target];
    }

    if (!isNaN(target)){
        return parseInt(target);
    }
    let instruction = instructions.find(x => x.output == target);
    
    let op1 = !isNaN(instruction.getInputs()[0]) ? parseInt(instruction.getInputs()[0]) : eval(instruction.getInputs()[0]);
    let op2 = 0;
    if (instruction.getInputs().length > 1){
        op2 = !isNaN(instruction.getInputs()[1]) ? parseInt(instruction.getInputs()[1]) : eval(instruction.getInputs()[1]);
    }

    let result = 0;
    switch(instruction.type){
        case "AND":
            result = op1 & op2;
            break;
        case "OR":
            result = op1 | op2;
            break;
        case "LSHIFT":
            result = op1 << op2;
            break;
        case "RSHIFT":
            result = op1 >> op2;
            break;
        case "NOT":
            result = ~op1;
            break;
        case "ASSIGN":
            result = op1;
            break;
    }

    hashTable[target] = result;
    return result;
}


main();